src/models.jl

```diff
@@ -0,0 +1,112 @@
+const default_jitter = 1e-6
+
+struct SVGPModel{Tlik}
+    kernel_func # function to construct the kernel from `k`
+    lik::Tlik   # the likelihood function
+    jitter      # the jitter added to covariance matrices
+    
+    ## Trainable parameters
+    k::AbstractVector           # kernel parameters
+    m::AbstractVector           # variational mean
+    A::AbstractMatrix           # variational covariance (sqrt)
+    z::AbstractVector           # inducing points
+end
+
+@functor SVGPModel (k, m, A, z,)
+
+function SVGPModel(
+    kernel_func,
+    kernel_params,
+    inducing_inputs;
+    q_μ::Union{AbstractVector, Nothing}=nothing,
+    q_Σ_sqrt::Union{AbstractMatrix, Nothing}=nothing,
+    jitter=default_jitter,
+    likelihood=GaussianLikelihood(jitter)
+)
+    m, A = _init_variational_params(q_μ, q_Σ_sqrt, inducing_inputs)
+    return SVGPModel(
+        kernel_func,
+        likelihood,
+        jitter,
+        kernel_params,
+        m,
+        A,
+        inducing_inputs
+    )
+end
+
+function (m::SVGPModel{<:GaussianLikelihood})(x)
+    f = prior(m)
+    fx = f(x, m.lik.σ²)
+    fu = f(m.z, m.jitter)
+    q = _construct_q(m)
+    return fx, fu, q
+end
+
+function (m::SVGPModel)(x)
+    f = prior(m)
+    fx = f(x)
+    fu = f(m.z).fx
+    q = _construct_q(m)
+    return fx, fu, q
+end
+
+function posterior(m::SVGPModel{<:GaussianLikelihood})
+    f = prior(m)
+    fu = f(m.z, m.jitter)
+    q = _construct_q(m)
+    return SparseGPs.approx_posterior(SVGP(), fu, q)
+end
+
+function posterior(m::SVGPModel)
+    f = prior(m)
+    fu = f(m.z).fx
+    q = _construct_q(m)
+    post = SparseGPs.approx_posterior(SVGP(), fu, q)
+    return LatentGP(post, m.lik, m.jitter) # TODO: should this return `post` instead?
+end
+
+function prior(m::SVGPModel{<:GaussianLikelihood})
+    kernel = m.kernel_func(m.k)
+    return GP(kernel)
+end
+
+function prior(m::SVGPModel)
+    kernel = m.kernel_func(m.k)
+    return LatentGP(GP(kernel), m.lik, m.jitter)
+end
+
+function loss(m::SVGPModel, x, y; n_data=length(y))
+    return -elbo(m, x, y; n_data)
+end
+
+function elbo(m::SVGPModel, x, y; n_data=length(y))
+    fx, fu, q = m(x)
+    return SparseGPs.elbo(fx, y, fu, q; n_data)
+end
+
+function _init_variational_params(n)
+    m = zeros(n)
+    A = Matrix{Float64}(I, n, n)
+    return m, A
+end
+
+function _init_variational_params(
```

Comment:
Should we add a keyword argument to enforce the eltype and type of \mu and \Sigma?

Reply:
I added an eltype parameter, but I can't think of a reason why you wouldn't want to use `Vector` / `Matrix`, so maybe a type parameter isn't needed?

Reply:
Well in the crazy case where someone would like to use GPUs, one would need pass CuArrays but I guess they would just have to make the effort of passing their own variational parameters.

Reply:
Since it's a functor, isn't it easiest to just use `Flux.gpu` for that after it's initialised?